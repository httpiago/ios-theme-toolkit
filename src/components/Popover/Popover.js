import React from 'react'
import PropTypes from 'prop-types'
import { mapClassList } from '../../Utils'
import Portal from '../Portal/Portal'

/**
 * Popover component.
 *
 * @since 0.2.0
 * @see https://httpiago.github.io/ios-theme-toolkit/#/components/Popover
 *
 * @todo Make more keyboard-friendly navigation
 */
export default class Popover extends React.Component {
  state = {
    isOpen: false,

    popoverPosition: { top: 0, left: 0},
    arrowPosition: { top: 0, left: 0}
  }

  childrenRef = React.createRef() // Referência ao elemento filho
  popoverRef = React.createRef() // Referência ao popover criado na tela

  componentDidMount() {
    if (this.props.trigger === 'hover') {
      var t = null
      this.childrenRef.current.addEventListener('mouseenter', this.me = () => {
        t = setTimeout(this.open, this.props.mouseEnterDelay)
      })
      this.childrenRef.current.addEventListener('mouseleave', this.ml = () => {
        clearTimeout(t)
        this.close()
      })
    } else {
      document.addEventListener('click', this.handleClick)
    }
  }
  componentWillUnmount() {
    document.removeEventListener('click', this.handleClick)
    this.childrenRef.current.removeEventListener('mouseenter', this.me)
    this.childrenRef.current.removeEventListener('mouseleave', this.ml)
  }

  /**
   * Mostrar popover na tela.
   */
  open = () => {
    this.setState({ isOpen: true }, this.calcPopoverPosition)

    this.popoverRef.current.getElementsByClassName('popover__content')[0].focus()
  }

  /**
   * Ocultar popover na tela.
   */
  close = () => {
    this.setState({ isOpen: false })
  }

  /**
   * Alterar visibilidade com base no estado anterior.
   */
  troggleVisibility = () => {
    let { isOpen } = this.state

    if (isOpen === true) this.close()
    else this.open()
  }

  /**
   * Calcular a posição do popover baseado no elemento filho.
   */
  calcPopoverPosition = () => {
    const child = this.childrenRef.current.getBoundingClientRect()
    const pop = this.popoverRef.current.getBoundingClientRect()

    const childWidthMiddle = child.left + (child.width / 2)
    const childHeightMiddle = child.top + (child.height / 2)
    const maxLeft = window.innerWidth - pop.width - 4 // Impedir que o popover ultrapasse os limites da tela

    switch (this.props.placement) {
      // Popover posicionado abaixo do elemento
      case 'bottom': {
        var popoverPosition = {
          top: (window.scrollY + child.top + child.height + 10),
          left: window.scrollX + Math.max(4, Math.min(maxLeft, (childWidthMiddle - (pop.width / 2))))
        }
        var arrowPosition = {
          top: -(13 / 2),
          left: childWidthMiddle - popoverPosition.left
        }
        break;
      }

      // Popover posicionado acima do elemento
      case 'top': {
        var popoverPosition = {
          top: (window.scrollY + child.top - pop.height - 5),
          left: window.scrollX + Math.max(4, Math.min(maxLeft, (childWidthMiddle - (pop.width / 2))))
        }
        var arrowPosition = {
          top: pop.height - (20 / 2),
          left: childWidthMiddle - popoverPosition.left
        }
        break;
      }

      // Popover posicionado do lado direito do elemento
      case 'right': {
        var popoverPosition = {
          top: (window.scrollY + childHeightMiddle - (pop.height / 2) + 2),
          left: (window.scrollX + child.left + child.width + 3)
        }
        var arrowPosition = {
          top: (pop.height / 2) - (14 / 2),
          left: 4
        }
        break;
      }

      // Popover posicionado do lado esquerdo do elemento
      case 'left': {
        var popoverPosition = {
          top: (window.scrollY + childHeightMiddle - (pop.height / 2) + 2),
          left: (window.scrollX + child.left - pop.width - 3)
        }
        var arrowPosition = {
          top: (pop.height / 2) - (14 / 2),
          left: pop.width
        }
        break;
      }

      default: break;
    }

    this.setState({
      popoverPosition,
      arrowPosition
    })
  }

  /**
   * Função que é chamada em qualquer clique do usuário e checa o que deve ser feito,
   * se deve abrir ou fechar o popover.
   */
  handleClick = (event) => {
    if (this.childrenRef.current.contains(event.target)) {
      // Click no this.childrenRef
      this.troggleVisibility()
    } else {
      // Click fora this.childrenRef

      // Checar se o click foi dentro do popover e se a propriedade "closeOnInnerClick" for true, fechar o popover
      if (
        this.popoverRef.current && this.popoverRef.current.contains(event.target)
        && !this.props.closeOnInnerClick
      ) return;

      this.close()
    }
  }

  render() {
    const { children, content, backdrop, placement, withoutArrow, color, className: aditionalClasses, trigger, closeOnInnerClick, mouseEnterDelay, ...rest } = this.props

    const prefix = 'popover'
    const classes = mapClassList({
      [`${prefix}--${placement}`]: placement,
      [`${prefix}--${color}`]: color,
    })

    const { isOpen, popoverPosition, arrowPosition } = this.state

    return (
      <React.Fragment>
        <span ref={this.childrenRef}>
          {children}
        </span>

        {isOpen && (
          <Portal>
            {backdrop && (<div className="popover__backdrop" />)}

            <div
              ref={this.popoverRef}
              className={`${prefix} ${classes} ${aditionalClasses}`}
              style={popoverPosition}
              {...rest}
            >
              {!withoutArrow && (<span className="popover__arrow" style={arrowPosition} />)}
              <div className="popover__content" tabIndex="-1">{content}</div>
            </div>
          </Portal>
        )}
      </React.Fragment>
    )
  }
}

Popover.defaultProps = {
  className: '',
  backdrop: false,
  trigger: 'click',
  mouseEnterDelay: 250,
  placement: 'bottom',
  closeOnInnerClick: false,
  color: 'default',
  withoutArrow: false
}

Popover.propTypes = {
  /** Aditional classes. */
  className: PropTypes.string,
  /** Element the will be used to trigger the popover */
  children: PropTypes.node.isRequired,
  /** The content of popover (anything that can be rendered by React) */
  content: PropTypes.node.isRequired,
  /** Show a transparent background behind the popover */
  backdrop: PropTypes.bool,
  /** The trigger */
  trigger: PropTypes.oneOf([ 'click', 'hover' ]),
  /**  */
  mouseEnterDelay: PropTypes.number,
  /** Popover position */
  placement: PropTypes.oneOf([ 'top', 'left', 'bottom', 'right']),
  /** Close when user click inside popover */
  closeOnInnerClick: PropTypes.bool,
  /** Popover color */
  color: PropTypes.oneOf([ 'default', 'blue', 'red', 'green', 'yellow', 'gray', 'black' ]),
}
