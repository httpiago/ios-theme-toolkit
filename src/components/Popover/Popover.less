@import '../../customizations.less';

// Default style
.popover {
  background-color: #FFF;
  box-shadow: 0 2px 6px rgba(0,0,0,0.26);
  border-radius: 6px;
  position: absolute;
  max-width: 98%;
  z-index: 10010;
}

.popover__arrow {
  background-color: #FFF;
  border-radius: 4px;
  width: 16px;
  height: 16px;
  position: absolute;
  margin-left: calc(-20px / 2);
  left: 50%;
  top: calc(-13px / 2);
  -webkit-transform: rotate(45deg);
  transform: rotate(45deg);
}

.popover__backdrop {
  background-color: rgba(0,0,0,0.6);
  position: fixed;
  height: 100%;
  width: 100%;
  top: 0px;
  left: 0px;
  z-index: 10000;
}

.popover__content {
  max-width: 100%;
  padding: 12px 15px;
  overflow: hidden;
  font-family: inherit;
  font-size: 15px;
  line-height: 20px;
  color: #333;
}

// Blue popover
.popover.popover--blue {
  background-color: @blue-color;

  & .popover__arrow {
    background-color: @blue-color;
  }

  & .popover__content {
    color: #FFFFFF;
  }
}

// Red popover
.popover.popover--red {
  background-color: @red-color;

  & .popover__arrow {
    background-color: @red-color;
  }

  & .popover__content {
    color: #FFFFFF;
  }
}

// Green popover
.popover.popover--green {
  background-color: @green-color;

  & .popover__arrow {
    background-color: @green-color;
  }

  & .popover__content {
    color: #FFFFFF;
  }
}

// Yellow popover
.popover.popover--yellow {
  background-color: @yellow-color;

  & .popover__arrow {
    background-color: @yellow-color;
  }

  & .popover__content {
    color: darken(@yellow-color, 40%);
  }
}

// Gray popover
.popover.popover--gray {
  background-color: @gray-color;

  & .popover__arrow {
    background-color: @gray-color;
  }

  & .popover__content {
    color: #333333;
  }
}

// Black popover
.popover.popover--black {
  background-color: @black-color;

  & .popover__arrow {
    background-color: @black-color;
  }

  & .popover__content {
    color: #FFFFFF;
  }
}


.popover__content > .button-group {
  margin: -12px -14px;
  min-width: 120px;
}
