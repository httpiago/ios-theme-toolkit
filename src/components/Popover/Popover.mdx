---
name: Popover
route: /components/Popover
menu: Components
---

import { Playground, PropsTable } from 'docz'
import { Popover, Button, ButtonGroup } from '../../index.js'
import '../../../dist/styles.css'

# Popover

To use this component, simply import it into any file and pass the [modifiers](#/#props) in the properties.

```jsx
import { Popover, Button } from 'ios-theme-toolkit'

const popoverContent = <span>Hello world</span>

export default () => (
  <Popover
    content={popoverContent}
  >
    <Button>Open popover</Button>
  </Popover>
)
```

## Examples
<Playground>
  <span className="desc">Default position (bottom):</span>
  <Popover content={(<span>Hello world</span>)}>
    <Button color="blue" fill>Open popover</Button>
  </Popover>

  <span className="desc">Top:</span>
  <Popover placement="top" content={(<span>Hello world</span>)}>
    <Button color="blue" fill>Open popover</Button>
  </Popover>

  <span className="desc">Left:</span>
  <Popover placement="left" content={(<span>Hello world</span>)}>
    <Button color="blue" fill>Open popover</Button>
  </Popover>

  <span className="desc">Right:</span>
  <Popover placement="right" content={(<span>Hello world</span>)}>
    <Button color="blue" fill>Open popover</Button>
  </Popover>

  <span className="desc">With backdrop:</span>
  <Popover backdrop={true} closeOnInnerClick={true} content={(<span>Hello world</span>)}>
    <Button color="blue" fill>Open popover</Button>
  </Popover>

  <span className="desc">Hover trigger:</span>
  <Popover trigger="hover" content={(<span>Hello world</span>)}>
    <Button color="blue" fill>Open popover</Button>
  </Popover>

  <span className="desc">Colorful:</span>
  <Popover color="blue" content={(<span>Hello world</span>)}><Button color="blue">Open popover</Button></Popover>
  <Popover color="red" content={(<span>Hello world</span>)}><Button color="red">Open popover</Button></Popover>
  <Popover color="green" content={(<span>Hello world</span>)}><Button color="green">Open popover</Button></Popover>
  <Popover color="yellow" content={(<span>Hello world</span>)}><Button color="yellow">Open popover</Button></Popover>
  <Popover color="gray" content={(<span>Hello world</span>)}><Button color="gray">Open popover</Button></Popover>
  <Popover color="black" content={(<span>Hello world</span>)}><Button color="black">Open popover</Button></Popover>

  <span className="desc">With buttons:</span>
  <Popover closeOnInnerClick={true} content={(
    <ButtonGroup direction="vertical" outline>
      <Button autoFocus>Option 1</Button>
      <Button>Option 2</Button>
      <Button>Option 3</Button>
      <Button>Option 4</Button>
    </ButtonGroup>
  )}>
    <Button color="blue" fill>Open popover</Button>
  </Popover>

  <style>{".react-live-preview .desc { display: block; font-weight: bold; margin-top: 20px }"}</style>
</Playground>

#### Tip

If you put some element that can receive focus (buttons, inputs ...) inside the popover, place an "autofocus" attribute on the element to keyboard-friendly navigation.

## Props

<PropsTable of={ Popover } />

All the properties which are not listed above will be transferred to the  \<div class="popover"\> tag.
