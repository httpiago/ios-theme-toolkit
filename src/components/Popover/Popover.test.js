/**
 * TEST FILE
 * Please try writing the tests for all component behaviors,
 * manually call the functions within the class, change the properties...
 *
 * @see https://jestjs.io/docs/en/expect
 * @see https://github.com/airbnb/enzyme/tree/master/docs/api
 */
import React from 'react'
import { shallow } from 'enzyme'
import Popover from './Popover'
import 'jsdom-global/register' // Create fake DOM

const COMPONENT = shallow(
  <Popover
    content={'Test content'}
    className="aditionalClass"
  >
    <div className="check-children">Hello world</div>
  </Popover>
)

describe('Test Popover component', () => {
  it('should render correctly based on last snapshot', () => {
    expect(COMPONENT).toMatchSnapshot()
  })

})
