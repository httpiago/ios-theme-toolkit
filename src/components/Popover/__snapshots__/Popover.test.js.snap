// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test Popover component should render correctly based on last snapshot 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Popover
    backdrop={false}
    className="aditionalClass"
    closeOnInnerClick={false}
    color="default"
    content="Test content"
    mouseEnterDelay={250}
    placement="bottom"
    trigger="click"
    withoutArrow={false}
  >
    <div
      className="check-children"
    >
      Hello world
    </div>
  </Popover>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "function",
    "props": Object {
      "children": Array [
        <span>
          <div
            className="check-children"
          >
            Hello world
          </div>
        </span>,
        false,
      ],
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": <div
            className="check-children"
          >
            Hello world
          </div>,
        },
        "ref": Object {
          "current": null,
        },
        "rendered": Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "Hello world",
            "className": "check-children",
          },
          "ref": null,
          "rendered": "Hello world",
          "type": "div",
        },
        "type": "span",
      },
      false,
    ],
    "type": Symbol(react.fragment),
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "function",
      "props": Object {
        "children": Array [
          <span>
            <div
              className="check-children"
            >
              Hello world
            </div>
          </span>,
          false,
        ],
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": <div
              className="check-children"
            >
              Hello world
            </div>,
          },
          "ref": Object {
            "current": null,
          },
          "rendered": Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": "Hello world",
              "className": "check-children",
            },
            "ref": null,
            "rendered": "Hello world",
            "type": "div",
          },
          "type": "span",
        },
        false,
      ],
      "type": Symbol(react.fragment),
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;
